cpp/src/arrow/io/file.cc

```diff
@@ -118,33 +118,78 @@ namespace io {
 
 // ----------------------------------------------------------------------
 // Cross-platform file compatability layer
+
 #if defined(_MSC_VER)
+
 constexpr const char* kRangeExceptionError =
     "Range exception during wide-char string conversion";
+
+class PlatformFilename {
```

Comment:
Is there any reason not to use the existing boost filesystem library? Does it not provide APIs that abstract over OS specific path details?